control-plane-operator/controllers/awsprivatelink/awsprivatelink_controller.go

```diff
@@ -0,0 +1,127 @@
+package awsprivatelink
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/informers"
+	kubeclient "k8s.io/client-go/kubernetes"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+	"github.com/openshift/hypershift/support/upsert"
+)
+
+const (
+	defaultResync = 10 * time.Hour
+)
+
+type PrivateServiceObserver struct {
+	client client.Client
+	log    logr.Logger
+
+	ControllerName   string
+	ServiceNamespace string
+	ServiceName      string
+	HCPNamespace     string
+	upsert.CreateOrUpdateProvider
+}
+
+func nameMapper(names []string) handler.MapFunc {
+	nameSet := sets.NewString(names...)
+	return func(obj client.Object) []reconcile.Request {
+		if !nameSet.Has(obj.GetName()) {
+			return nil
+		}
+		return []reconcile.Request{
+			{
+				NamespacedName: types.NamespacedName{
+					Namespace: obj.GetNamespace(),
+					Name:      obj.GetName(),
+				},
+			},
+		}
+	}
+}
+
+func namedResourceHandler(names ...string) handler.EventHandler {
+	return handler.EnqueueRequestsFromMapFunc(nameMapper(names))
+}
+
+func ControllerName(name string) string {
+	return fmt.Sprintf("%s-observer", name)
+}
+
+func (r *PrivateServiceObserver) SetupWithManager(ctx context.Context, mgr ctrl.Manager) error {
+	r.log = ctrl.Log.WithName(r.ControllerName).WithValues("name", r.ServiceName, "namespace", r.ServiceNamespace)
```

Comment:
Nit: can use `log := ctrl.LoggerFrom(ctx).WithValues()`

Reply:
But then we override it in every iteration?

Reply:
In this case, `ctx` is the context in the main function setting up the controllers.  Not sure that is what we want.